File: rust/fory-core/src/serializer/string.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::error::Error;
use crate::meta::get_latin1_length;
use crate::resolver::context::ReadContext;
use crate::resolver::context::WriteContext;
use crate::resolver::type_resolver::TypeResolver;
use crate::serializer::util::read_basic_type_info;
use crate::serializer::{ForyDefault, Serializer};
use crate::types::TypeId;
use std::mem;

enum StrEncoding {
    Latin1 = 0,
    Utf16 = 1,
    Utf8 = 2,
}

impl Serializer for String {
    #[inline(always)]
    fn fory_write_data(&self, context: &mut WriteContext) -> Result<(), Error> {
        if !context.is_xlang() {
            // Fast path: non-xlang mode always uses UTF-8 without encoding header
            context.writer.write_varuint32(self.len() as u32);
            context.writer.write_utf8_string(self);
            return Ok(());
        }

        // xlang mode: use encoding header for optimal format selection
        let mut len = get_latin1_length(self);
        if len >= 0 {
            let bitor = (len as u64) << 2 | StrEncoding::Latin1 as u64;
            context.writer.write_varuint36_small(bitor);
            context.writer.write_latin1_string(self);
        } else if context.is_compress_string() {
            // todo: support `writeNumUtf16BytesForUtf8Encoding` like in java
            len = self.len() as i32;
            let bitor = (len as u64) << 2 | StrEncoding::Utf8 as u64;
            context.writer.write_varuint36_small(bitor);
            context.writer.write_utf8_string(self);
        } else {
            let utf16: Vec<u16> = self.encode_utf16().collect();
            let bitor = (utf16.len() as u64 * 2) << 2 | StrEncoding::Utf16 as u64;
            context.writer.write_varuint36_small(bitor);
            context.writer.write_utf16_bytes(&utf16);
        }
        Ok(())
    }

    #[inline(always)]
    fn fory_read_data(context: &mut ReadContext) -> Result<Self, Error> {
        if !context.is_xlang() {
            // Fast path: non-xlang mode always uses UTF-8 without encoding header
            let len = context.reader.read_varuint32()? as usize;
            return context.reader.read_utf8_string(len);
        }

        // xlang mode: read encoding header and decode accordingly
        let bitor = context.reader.read_varuint36small()?;
        let len = bitor >> 2;
        let encoding = bitor & 0b11;
        let encoding = match encoding {
            0 => StrEncoding::Latin1,
            1 => StrEncoding::Utf16,
            2 => StrEncoding::Utf8,
            _ => {
                return Err(Error::encoding_error(format!(
                    "wrong encoding value: {}",
                    encoding
                )))
            }
        };
        let s = match encoding {
            StrEncoding::Latin1 => context.reader.read_latin1_string(len as usize),
            StrEncoding::Utf16 => context.reader.read_utf16_string(len as usize),
            StrEncoding::Utf8 => context.reader.read_utf8_string(len as usize),
        }?;
        Ok(s)
    }

    #[inline(always)]
    fn fory_reserved_space() -> usize {
        mem::size_of::<i32>()
    }

    #[inline(always)]
    fn fory_get_type_id(_: &TypeResolver) -> Result<u32, Error> {
        Ok(TypeId::STRING as u32)
    }

    #[inline(always)]
    fn fory_type_id_dyn(&self, _: &TypeResolver) -> Result<u32, Error> {
        Ok(TypeId::STRING as u32)
    }

    #[inline(always)]
    fn fory_static_type_id() -> TypeId
    where
        Self: Sized,
    {
        TypeId::STRING
    }

    #[inline(always)]
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    #[inline(always)]
    fn fory_write_type_info(context: &mut WriteContext) -> Result<(), Error> {
        context.writer.write_varuint32(TypeId::STRING as u32);
        Ok(())
    }

    #[inline(always)]
    fn fory_read_type_info(context: &mut ReadContext) -> Result<(), Error> {
        read_basic_type_info::<Self>(context)
    }
}

impl ForyDefault for String {
    #[inline(always)]
    fn fory_default() -> Self {
        String::new()
    }
}

File: go/fory/string.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fory

import (
	"fmt"
	"reflect"
	"unicode/utf16"
)

// Encoding type constants
const (
	encodingLatin1  = iota // Latin1/ISO-8859-1 encoding
	encodingUTF16LE        // UTF-16 Little Endian encoding
	encodingUTF8           // UTF-8 encoding (default)
)

// writeString writes a string to buffer using xlang encoding
// Optimized to combine header and data writes with single grow() call
func writeString(buf *ByteBuffer, value string) {
	data := unsafeGetBytes(value)
	dataLen := len(data)
	header := (uint64(dataLen) << 2) | encodingUTF8

	// Reserve space for header (max 5 bytes) + data in one call
	buf.Reserve(5 + dataLen)
	// Write header inline without grow check
	buf.UnsafeWriteVaruint64(header)
	// Write data inline without grow check
	if dataLen > 0 {
		copy(buf.data[buf.writerIndex:], data)
		buf.writerIndex += dataLen
	}
}

// readString reads a string from buffer using xlang encoding
func readString(buf *ByteBuffer, err *Error) string {
	header := buf.ReadVaruint36Small(err)
	size := header >> 2       // Extract byte count
	encoding := header & 0b11 // Extract encoding type

	switch encoding {
	case encodingLatin1:
		return readLatin1(buf, int(size), err)
	case encodingUTF16LE:
		// For UTF16LE, size is byte count, need to convert to char count
		return readUTF16LE(buf, int(size), err)
	case encodingUTF8:
		return readUTF8(buf, int(size), err)
	default:
		err.SetError(fmt.Errorf("invalid string encoding: %d", encoding))
		return ""
	}
}

// Specific encoding read methods
func readLatin1(buf *ByteBuffer, size int, err *Error) string {
	data := buf.ReadBinary(size, err)
	// Latin1 bytes need to be converted to UTF-8
	// Each Latin1 byte is a single Unicode code point (0-255)
	runes := make([]rune, size)
	for i, b := range data {
		runes[i] = rune(b)
	}
	return string(runes)
}

func readUTF16LE(buf *ByteBuffer, byteCount int, err *Error) string {
	data := buf.ReadBinary(byteCount, err)

	// Reconstruct UTF-16 code units
	charCount := byteCount / 2
	u16s := make([]uint16, charCount)
	for i := 0; i < byteCount; i += 2 {
		u16s[i/2] = uint16(data[i]) | uint16(data[i+1])<<8
	}

	return string(utf16.Decode(u16s))
}

func readUTF8(buf *ByteBuffer, size int, err *Error) string {
	data := buf.ReadBinary(size, err)
	return string(data) // Direct UTF-8 conversion
}

// ============================================================================
// String Serializers - implement unified Serializer interface
// ============================================================================

// stringSerializer handles string type
type stringSerializer struct{}

var globalStringSerializer = stringSerializer{}

func (s stringSerializer) WriteData(ctx *WriteContext, value reflect.Value) {
	writeString(ctx.buffer, value.String())
}

func (s stringSerializer) Write(ctx *WriteContext, refMode RefMode, writeType bool, hasGenerics bool, value reflect.Value) {
	if refMode != RefModeNone {
		// String is non-primitive, needs ref flag
		ctx.buffer.WriteInt8(NotNullValueFlag)
	}
	if writeType {
		ctx.buffer.WriteVaruint32Small7(uint32(STRING))
	}
	s.WriteData(ctx, value)
}

func (s stringSerializer) ReadData(ctx *ReadContext, value reflect.Value) {
	err := ctx.Err()
	str := readString(ctx.buffer, err)
	if ctx.HasError() {
		return
	}
	value.SetString(str)
}

func (s stringSerializer) Read(ctx *ReadContext, refMode RefMode, readType bool, hasGenerics bool, value reflect.Value) {
	err := ctx.Err()
	if refMode != RefModeNone {
		// String is non-primitive, needs ref flag
		refFlag := ctx.buffer.ReadInt8(err)
		if refFlag == NullFlag {
			value.SetString("")
			return
		}
	}
	if readType {
		_ = ctx.buffer.ReadVaruint32Small7(err)
	}
	if ctx.HasError() {
		return
	}
	s.ReadData(ctx, value)
}

func (s stringSerializer) ReadWithTypeInfo(ctx *ReadContext, refMode RefMode, typeInfo *TypeInfo, value reflect.Value) {
	s.Read(ctx, refMode, false, false, value)
}

// ptrToStringSerializer serializes a pointer to string
type ptrToStringSerializer struct{}

func (s ptrToStringSerializer) Write(ctx *WriteContext, refMode RefMode, writeType bool, hasGenerics bool, value reflect.Value) {
	if refMode != RefModeNone {
		if value.IsNil() {
			ctx.buffer.WriteInt8(NullFlag)
			return
		}
		ctx.buffer.WriteInt8(NotNullValueFlag)
	}
	if writeType {
		ctx.buffer.WriteVaruint32Small7(uint32(STRING))
	}
	s.WriteData(ctx, value)
}

func (s ptrToStringSerializer) WriteData(ctx *WriteContext, value reflect.Value) {
	str := value.Interface().(*string)
	writeString(ctx.buffer, *str)
}

func (s ptrToStringSerializer) Read(ctx *ReadContext, refMode RefMode, readType bool, hasGenerics bool, value reflect.Value) {
	err := ctx.Err()
	if refMode != RefModeNone {
		if ctx.buffer.ReadInt8(err) == NullFlag {
			return
		}
	}
	if readType {
		_ = ctx.buffer.ReadVaruint32Small7(err)
	}
	if ctx.HasError() {
		return
	}
	s.ReadData(ctx, value)
}

func (s ptrToStringSerializer) ReadData(ctx *ReadContext, value reflect.Value) {
	err := ctx.Err()
	str := readString(ctx.buffer, err)
	if ctx.HasError() {
		return
	}
	ptr := new(string)
	*ptr = str
	value.Set(reflect.ValueOf(ptr))
}

func (s ptrToStringSerializer) ReadWithTypeInfo(ctx *ReadContext, refMode RefMode, typeInfo *TypeInfo, value reflect.Value) {
	s.Read(ctx, refMode, false, false, value)
}
